domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,310 @@ class TopicListController @Inject constructor(
       .build()
   }
 
-  private fun createOngoingStoryListFromProgress(
+  private fun createUpcomingTopicSummaryFromJson(
+    topicId: String,
+    jsonObject: JSONObject
+  ): UpcomingTopic {
+    val upcomingTopic = UpcomingTopic.newBuilder()
+    var totalChapterCount = 0
+    val storyData = jsonObject.getJSONArray("canonical_story_dicts")
+    for (i in 0 until storyData.length()) {
+      totalChapterCount += storyData
+        .getJSONObject(i)
+        .getJSONArray("node_titles")
+        .length()
+    }
+    val topicPlayAvailability = if (jsonObject.getBoolean("published")) {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayNow(true).build()
+    } else {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayInFuture(true).build()
+    }
+
+    return upcomingTopic.setTopicId(topicId)
+      .setName(jsonObject.getString("topic_name"))
+      .setVersion(jsonObject.optInt("version"))
+      .setTopicPlayAvailability(topicPlayAvailability)
+      .setLessonThumbnail(createTopicThumbnail(jsonObject))
+      .build()
+  }
+
+  private fun computePromotedActivityList(
+    topicProgressList: List<TopicProgress>
+  ): PromotedActivityList {
+    val promotedActivityListBuilder = PromotedActivityList.newBuilder()
+    if (topicProgressList.isNotEmpty()) {
+      promotedActivityListBuilder.promotedStoryList = computePromotedStoryList(topicProgressList)
+      if (promotedActivityListBuilder.promotedStoryList.getTotalPromotedStoryCount() == 0) {
+        promotedActivityListBuilder.comingSoonTopicList = computeComingSoonTopicList()
+      }
+    }
+    return promotedActivityListBuilder.build()
+  }
+
+  private fun computePromotedStoryList(topicProgressList: List<TopicProgress>): PromotedStoryList {
+    val promotedStoryListBuilder = PromotedStoryList.newBuilder()
```

Comment:
Optional nit
This can be future reduced to 
```
    return PromotedStoryList.newBuilder()
      .addAllUpTo(
        populateRecentlyPlayedStories(topicProgressList),
        PromotedStoryList.Builder::addAllRecentlyPlayedStory,
        limit = 3
      )
      .addAllUpTo(
        populateOlderPlayedStories(topicProgressList),
        PromotedStoryList.Builder::addAllOlderPlayedStory,
        limit = 3
      )
      .addAllUpTo(
        computeSuggestedStories(topicProgressList),
        PromotedStoryList.Builder::addAllSuggestedStory,
        limit = 3
      )
      .build()
```

Reply:
Done. @rt4914 I also need one suggestion here. For Tablet View we are showing 4 items, so setting limit to 3 over here is correct?

Reply:
https://github.com/oppia/oppia-android/pull/2272#discussion_r560727940